feat: support server-side i18n integration

kazupon opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxtjs.com/docs/community/contribution
-->

### 🔗 Linked issue

<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR is feature PR of `@intlify/h3` integration for nuxt i18n, and **experimental feature**.
I will also add docs.

we will consider some stuff the following:

- locale messages (i18n resources) will be able to use nuxt i18n options as much as possible
  - set the `defineI18nMiddleware` in the initialization of the nitro plugin
  - but, setting up all locale messages in the nitro plugin can have a performance impact, so consider lazy loading (we should load the locale when it is detected by the locale detector).
- server-side locale detection is done by a module that exports a user-defined locale detector function to the `experimental.localeDetector` option in the nuxt i18n config.
  - see about detauls for how to customize for locale detection: https://github.com/intlify/h3#%EF%B8%8F-custom-locale-detection
  - Ideally, it would be useful if locale detection could be seamlessly integrated with the nuxt i18n routing strategy and detect browserlanguage option. but, we don’t know use cases of server-side i18n, the first implementation allows users to customize locale detection on their own.


<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

kazupon committed:
feat: support server-side i18n integration

kazupon committed:
fix: locale detector loading on server-side

kazupon commented:
This PR is still WIP.

kazupon committed:
refactor

kazupon committed:
merge from main branch

kazupon committed:
fix: more deps

kazupon committed:
fix: cannot bundle intlify/core

kazupon committed:
fix: initial options loading

kazupon committed:
refactor: function name

kazupon committed:
Merge branch 'main' into feat/intlify-h3-integration2

kazupon committed:
refactor: remove context

kazupon committed:
fix: type errors

kazupon committed:
refactor: argument ordering

kazupon committed:
fix: pass user cache messages

kazupon committed:
fix: wrong judgement enabling

kazupon committed:
fix: not work useRuntimeConfig

kazupon committed:
Merge branch 'main' into feat/intlify-h3-integration3

BobbieGoede committed:
feat: `useRuntimeConfig` support (#2572)

* feat: hacky runtime config support

* refactor: remove unnecessary function passing

* refactor: simplify and make typesafe

kazupon committed:
fix: spec tests

kazupon committed:
chore: bump nitropack

BobbieGoede commented on specs/basic_usage.spec.ts:
```diff
@@ -97,8 +107,8 @@ test('layer vueI18n options properties are merge and override by priority', asyn
-  expect(await getText(page, '#snake-case')).toEqual('À-propos-de-ce-site')
-  expect(await getText(page, '#pascal-case')).toEqual('ÀProposDeCeSite')
+  expect(await getText(page, '#snake-case')).toEqual('Should-be-overridden')
+  expect(await getText(page, '#pascal-case')).toEqual('ShouldBeOverridden')
```
```suggestion
  // Project VueI18n config overwrites `Should-be-overridden` and `ShouldBeOverridden`
  expect(await getText(page, '#snake-case')).toEqual('À-propos-de-ce-site')
  expect(await getText(page, '#pascal-case')).toEqual('ÀProposDeCeSite')
```

I know this is still WIP, but this should be reverted otherwise you may be testing incorrect behavior 😅

kazupon replied:
Thanks!
I've reverted these assetions.
Fixed in https://github.com/nuxt-modules/i18n/pull/2558/commits/01bb141b75468454d746220d0594c3d46bf15ed4

kazupon committed:
fix: salvage basic_usage.spec.ts

kazupon committed:
fix: Avoid order broken of `vueI18nConfigPaths`

kazupon committed:
fix: type error

kazupon committed:
refactor: move to messages

kazupon committed:
fix: lazy load implementation

kazupon committed:
fix: i18n resources always lazy loading for nitro side

kazupon committed:
fix: reduce resource size for nitropack bundling

kazupon committed:
fix: drop import assetion

NOTE:
move to import attribute.
ref: https://github.com/tc39/proposal-import-attributes
In the first place, there is no point in using it, since general import can also resolve too.

kazupon committed:
refactor

BobbieGoede committed:
fix: h3 integration layer support (#2589)

* fix: server integration layer support

* test: test `localeDetector` usage

kazupon committed:
Merge branch 'main' into feat/intlify-h3-integration

kazupon committed:
chore: update deps

kazupon committed:
chore: upgrade vitest

kazupon committed:
fix: cannot resolve type for intlify/h3 utilites

kazupon committed:
fix: bump intlify/h3 and utilts

kazupon committed:
fix: tweak locale detector for playground

kazupon committed:
refactor

kazupon committed:
add `defineI18nLocaleDetector` composable

kazupon committed:
docs: add `useTranslation`

kazupon committed:
fix: remove console log

kazupon committed:
docs: update server-side translations

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
```diff
@@ -0,0 +1,91 @@
+# Server-side Translations
+
+The locale mesasges defined in nuxt i18n module options are integrated, so all you need to do is configure the locale detector. 
```
```suggestion
The locale messages defined in nuxt i18n module options are integrated, so all you need to do is configure the locale detector. 
```

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
```diff
@@ -0,0 +1,91 @@
+---
+
+::alert{type="warning"}
+**This feature is experimantal,** that is supported from v8 RC8.
```
```suggestion
**This feature is experimental,** that is supported from v8 RC8.
```

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
```diff
@@ -0,0 +1,91 @@
+
+For server-side translation, you need to define locale detector.
+
+nuxt i18n export the `defineI18nLocaleDetector` composable function to define it.
```
```suggestion
Nuxt i18n export the `defineI18nLocaleDetector` composable function to define it.
```

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
```diff
@@ -0,0 +1,91 @@
+
+when you will define the locale detector, you need to pass the path to the locale detector to `experimental.localeDetector` option.
+
+The following is an example of locale detector configuration defined directly under nuxt applicatoin:
```
```suggestion
The following is an example of locale detector configuration defined directly under Nuxt application:
```

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
````diff
@@ -0,0 +1,91 @@
+})
+```
+
+The locale detector fucntion is used to detect the locale on server-side. It's called per request on the server.
````
```suggestion
The locale detector function is used to detect the locale on server-side. It's called per request on the server.
```

BobbieGoede commented on docs/content/2.guide/16.server-side-translations.md:
````diff
@@ -0,0 +1,91 @@
+```ts
+// you need to define `async` event handler
+export default defineEventHandler(async event => {
+  // call `useTranslation`, so it retrun the translation function
````
```suggestion
  // call `useTranslation`, which returns the translation function
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
````diff
@@ -177,6 +177,32 @@ Note that if the value of `detectBrowserLanguage.useCookie` is `false`, an **emp
+})
+```
+
+The locale is use by the translation function is the locale detected by the function defined in [`experimental.localeDetector` option](../options/misc#experimental).
````
```suggestion
The translation function uses the locale detected by the function defined in [`experimental.localeDetector`](../options/misc#experimental).
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
```diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+
+The `defineI18nLocaleDetector` defines a composable function to detect the locale on the server-side.
+
+The locale detector fucntion is used to detect the locale on server-side. It's called per request on the server.
```
```suggestion
The locale detector function is used to detect the locale on the server-side, it is called per request on the server.
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
```diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+
+The locale detector fucntion is used to detect the locale on server-side. It's called per request on the server.
+
+You need return locale string.
```
```suggestion
The locale detector must return a locale string.
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
````diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+declare function defineI18nLocaleDetector(detector: (event: H3Event, config: LocaleConfig) => string): (event: H3Event, config: LocaleConfig) => string;
+```
+
+for locale detecting example:
````
```suggestion
An example of a locale detector:
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
````diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+})
+```
+
+In the locale detector function, you can use [`@intlify/h3` utilites](https://github.com/intlify/h3#%EF%B8%8F-utilites--helpers), these is auto imported.
````
```suggestion
In the locale detector function, you can use [`@intlify/h3` utilites](https://github.com/intlify/h3#%EF%B8%8F-utilites--helpers), these are auto imported.
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
```diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+
+  **Type**: `H3Event`
+
+  A H3 event. see details [H3 API docs](https://www.jsdocs.io/package/h3#H3Event)
```
```suggestion
  An H3 event. see details [H3 API docs](https://www.jsdocs.io/package/h3#H3Event)
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
```diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+
+  A locale config that is passed from Nitro. That has the below values:
+
+  - `defaultLocale`: This value is set by `defaultLocale` option of nuxt i18n option. if it's not set, that is set by `locale` that is loaded with `vueI18n` option (i18n.config). If these do not resolve the locale, use as default `'en-US'`.
```
```suggestion
  - `defaultLocale`: This value is set to the `defaultLocale` option of Nuxt i18n. If unset, it is set to the `locale` option loaded from the Vue I18n configuration (`i18n.config` file set on the `vueI18n` option). If neither of these are set, the default value of `'en-US'` is used.
```

BobbieGoede commented on docs/content/4.API/1.composables.md:
```diff
@@ -253,3 +279,73 @@ A function that is the dynamic locale messages loading, that has the following p
+
+  - `defaultLocale`: This value is set by `defaultLocale` option of nuxt i18n option. if it's not set, that is set by `locale` that is loaded with `vueI18n` option (i18n.config). If these do not resolve the locale, use as default `'en-US'`.
+
+  - `fallbackLocale`: This value is set by `fallbackLocale` option that is loaded with `vueI18n` option (i18n.config). If do not resolve the fallback locale, use as default `false`
```
```suggestion
  - `fallbackLocale`: This value is set to the `fallbackLocale` option loaded from the Vue I18n configuration (`i18n.config` file set on the `vueI18n` option). If no fallback locale has been configured this will default to `false`.
```

BobbieGoede reviewed (changes requested):
Nice, it looks great!

Only some small fixes/changes for the docs, other than that it looks like it can be merged 💪

kazupon committed:
Update docs/content/2.guide/16.server-side-translations.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/2.guide/16.server-side-translations.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/2.guide/16.server-side-translations.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/2.guide/16.server-side-translations.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/4.API/1.composables.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/4.API/1.composables.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/4.API/1.composables.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/4.API/1.composables.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon committed:
Update docs/content/4.API/1.composables.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>

kazupon commented:
@BobbieGoede 
Thanks for your review!
I'll merge this PR!

kazupon committed:
Update docs/content/2.guide/16.server-side-translations.md

Co-authored-by: Bobbie Goede <bobbiegoede@gmail.com>
